File: src/test/java/leets/weeth/domain/user/domain/repository/UserRepositoryTest.java
package leets.weeth.domain.user.domain.repository;

import static org.assertj.core.api.AssertionsForInterfaceTypes.*;

import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.context.annotation.Import;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;

import leets.weeth.config.TestContainersConfig;
import leets.weeth.domain.user.domain.entity.Cardinal;
import leets.weeth.domain.user.domain.entity.User;
import leets.weeth.domain.user.domain.entity.UserCardinal;
import leets.weeth.domain.user.domain.entity.enums.Status;
import leets.weeth.domain.user.test.fixture.CardinalTestFixture;
import leets.weeth.domain.user.test.fixture.UserCardinalTestFixture;
import leets.weeth.domain.user.test.fixture.UserTestFixture;

@DataJpaTest
@Import(TestContainersConfig.class)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
public class UserRepositoryTest {

	@Autowired
	private UserRepository userRepository;

	@Autowired
	private UserCardinalRepository userCardinalRepository;

	@Autowired
	private CardinalRepository cardinalRepository;

	private Cardinal cardinal7;
	private Cardinal cardinal8;

	@BeforeEach
	void setUp() {

		cardinal7 = cardinalRepository.save(CardinalTestFixture.createCardinal(7, 2026, 1));
		cardinal8 = cardinalRepository.save(CardinalTestFixture.createCardinal(8, 2026, 2));

		var user1 = userRepository.save(UserTestFixture.createActiveUser1());
		var user2 = userRepository.save(UserTestFixture.createActiveUser2());
		var user3 = userRepository.save(UserTestFixture.createWaitingUser1());

		user1.accept();
		user2.accept();
		userCardinalRepository.flush();

		userCardinalRepository.save(UserCardinalTestFixture.linkUserCardinal(user1, cardinal7));
		userCardinalRepository.save(UserCardinalTestFixture.linkUserCardinal(user2, cardinal8));
		userCardinalRepository.save(UserCardinalTestFixture.linkUserCardinal(user3, cardinal7));

	}

	@Test
	@DisplayName("findAllByCardinalAndStatus(): 특정 기수 + 상태에 맞는 유저만 조회된다")
	void findAllByCardinalAndStatus() {
		// when
		List<User> result = userRepository.findAllByCardinalAndStatus(cardinal7, Status.ACTIVE);

		// then
		assertThat(result)
			.hasSize(1)
			.extracting(User::getName)
			.containsExactly("적순");
	}

	@Test
	@DisplayName("findAllByStatusOrderByCardinalAndName() : 상태별로 최신 기수순 + 이름 오름차순으로 정렬된다")
	void findAllByStatusOrderByCardinalAndName() {
		//given
		Pageable pageable = PageRequest.of(0,10);

		//when
		Slice<User> resultSlice = userRepository.findAllByStatusOrderedByCardinalAndName(Status.ACTIVE, pageable);
		List<User> result = resultSlice.getContent();

		//then
		assertThat(result)
			.hasSize(2)
			.extracting(User::getName)
			.containsExactly("적순2", "적순");
	}

	@Test
	@DisplayName("findAllByCardinalOrderByNameAsc() : Active인 유저들 중 특정 기수 + 이름 오름차순으로 정렬한다.")
	void findAllByCardinalOrderByNameAsc() {
		//given
		Pageable pageable = PageRequest.of(0,10);

		//when
		Slice<User> resultSlice = userRepository.findAllByCardinalOrderByNameAsc(Status.ACTIVE, cardinal7,pageable);
		List<User> result = resultSlice.getContent();

		//then
		assertThat(result)
			.hasSize(1)
			.extracting(User::getName)
			.containsExactly("적순");
	}

}

File: src/test/kotlin/com/weeth/domain/user/domain/repository/UserRepositoryTest.kt
package com.weeth.domain.user.domain.repository

import com.weeth.config.TestContainersConfig
import com.weeth.domain.user.domain.entity.enums.Status
import com.weeth.domain.user.fixture.CardinalTestFixture
import com.weeth.domain.user.fixture.UserCardinalTestFixture
import com.weeth.domain.user.fixture.UserTestFixture
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.collections.shouldHaveSize
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.context.annotation.Import
import org.springframework.data.domain.PageRequest

@DataJpaTest
@Import(TestContainersConfig::class)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
class UserRepositoryTest(
    private val userRepository: UserRepository,
    private val userCardinalRepository: UserCardinalRepository,
    private val cardinalRepository: CardinalRepository,
) : DescribeSpec({

        lateinit var cardinal7: com.weeth.domain.user.domain.entity.Cardinal
        lateinit var cardinal8: com.weeth.domain.user.domain.entity.Cardinal

        beforeEach {
            cardinal7 = cardinalRepository.save(CardinalTestFixture.createCardinal(cardinalNumber = 7, year = 2026, semester = 1))
            cardinal8 = cardinalRepository.save(CardinalTestFixture.createCardinal(cardinalNumber = 8, year = 2026, semester = 2))

            val user1 = userRepository.save(UserTestFixture.createActiveUser1())
            val user2 = userRepository.save(UserTestFixture.createActiveUser2())
            val user3 = userRepository.save(UserTestFixture.createWaitingUser1())

            user1.accept()
            user2.accept()
            userCardinalRepository.flush()

            userCardinalRepository.save(UserCardinalTestFixture.linkUserCardinal(user1, cardinal7))
            userCardinalRepository.save(UserCardinalTestFixture.linkUserCardinal(user2, cardinal8))
            userCardinalRepository.save(UserCardinalTestFixture.linkUserCardinal(user3, cardinal7))
        }

        describe("findAllByCardinalAndStatus") {
            it("특정 기수 + 상태에 맞는 유저만 조회된다") {
                val result = userRepository.findAllByCardinalAndStatus(cardinal7, Status.ACTIVE)

                result shouldHaveSize 1
                result.map { it.name } shouldContainExactly listOf("적순")
            }
        }

        describe("findAllByStatusOrderedByCardinalAndName") {
            it("상태별로 최신 기수순 + 이름 오름차순으로 정렬된다") {
                val pageable = PageRequest.of(0, 10)

                val resultSlice = userRepository.findAllByStatusOrderedByCardinalAndName(Status.ACTIVE, pageable)
                val result = resultSlice.content

                result shouldHaveSize 2
                result.map { it.name } shouldContainExactly listOf("적순2", "적순")
            }
        }

        describe("findAllByCardinalOrderByNameAsc") {
            it("Active인 유저들 중 특정 기수 + 이름 오름차순으로 정렬한다") {
                val pageable = PageRequest.of(0, 10)

                val resultSlice = userRepository.findAllByCardinalOrderByNameAsc(Status.ACTIVE, cardinal7, pageable)
                val result = resultSlice.content

                result shouldHaveSize 1
                result.map { it.name } shouldContainExactly listOf("적순")
            }
        }
    })
